akka-stream/src/main/scala/akka/stream/scaladsl/Framing.scala

```diff
@@ -250,12 +277,12 @@ object Framing {
         @tailrec
         private def searchIndices(): Unit = {
           // Next possible position for the delimiter
-          val possibleMatchPos = buffer.indexOf(firstSeparatorByte, from = nextPossibleMatch)
+          val possibleMatchPos = buffer.indexWhere(delimitersByFirstByte.contains, from = nextPossibleMatch)
```

Comment:
I think this could have a significant performance impact on the single delimiter use case which is most important so we'd need to look at benchmark results and potentially special case a single separator back to `indexOf`. Also not sure that `indexWhere` is a good idea for the multi-delimiters, something that is cheaper to evaluate (a trie perhaps) may be needed.

You can find an existing jmh benchmark in `akka.stream.FramingBenchmark` in the akka-bench-jmh submodule. It can be run from sbt with something like `akka-bench-jmh/jmh:run -f 1 -w 5 -i 10 akka.stream.FramingBenchmark`, may also be good to run with `-prof gc` to look at the memory allocation pressure. Run on master and this PR and compare numbers (or report back as comments in the pr if you are unsure)

Reply:
@johanandren 
I didn’t manage to run benchmark...
I have the following errors:
[error] bnd: Missing file on classpath: /Users/mykhailoyeromenko/Documents/work/akka/akka-stream/target/scala-2.12/CompileJdk9-classes
and after I see a lot of:
[info] # Warmup Iteration   1: 2020-07-27 12:14:27,968 INFO  akka.event.slf4j.Slf4jLogger  - Slf4jLogger started MDC: {}
[info] <failure>
[info] java.lang.NullPointerException
[info] 	at akka.stream.FramingBenchmark.framing(FramingBenchmark.scala:88)
[info] 	at akka.stream.generated.FramingBenchmark_framing_jmhTest.framing_thrpt_jmhStub(FramingBenchmark_framing_jmhTest.java:143)
[info] 	at akka.stream.generated.FramingBenchmark_framing_jmhTest.framing_Throughput(FramingBenchmark_framing_jmhTest.java:83)
[info] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info] 	at java.lang.reflect.Method.invoke(Method.java:498)
[info] 	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:453)
[info] 	at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:437)
[info] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info] 	at java.lang.Thread.run(Thread.java:748)
[info] <JMH had finished, but forked VM did not exit, are there stray running threads? Waiting 24 seconds more...>
[info] Non-finished threads:
[info] Thread[test-akka.actor.default-dispatcher-5,5,main]
[info]   at sun.misc.Unsafe.park(Native Method)
[info]   at java.util.concurrent.ForkJoinPool.awaitWork(ForkJoinPool.java:1824)
[info]   at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1693)
[info]   at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:163)
[info] Thread[test-akka.actor.internal-dispatcher-4,5,main]
[info]   at sun.misc.Unsafe.park(Native Method)
[info]   at java.util.concurrent.ForkJoinPool.awaitWork(ForkJoinPool.java:1824)
[info]   at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1693)
[info]   at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:163)
[info] Thread[test-akka.actor.internal-dispatcher-2,5,main]
[info]   at sun.misc.Unsafe.park(Native Method)
[info]   at java.util.concurrent.ForkJoinPool.awaitWork(ForkJoinPool.java:1824)
[info]   at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1693)
[info]   at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:163)
[info] Thread[DestroyJavaVM,5,main]
[info] Thread[test-akka.actor.internal-dispatcher-3,5,main]
[info]   at sun.misc.Unsafe.park(Native Method)
[info]   at java.util.concurrent.ForkJoinPool.awaitWork(ForkJoinPool.java:1824)
[info]   at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1693)
[info]   at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:163)
[info] Thread[test-scheduler-1,5,main]
[info]   at java.lang.Thread.sleep(Native Method)
[info]   at akka.actor.LightArrayRevolverScheduler.waitNanos(LightArrayRevolverScheduler.scala:90)
[info]   at akka.actor.LightArrayRevolverScheduler$$anon$3.nextTick(LightArrayRevolverScheduler.scala:295)
[info]   at akka.actor.LightArrayRevolverScheduler$$anon$3.run(LightArrayRevolverScheduler.scala:265)
[info]   at java.lang.Thread.run(Thread.java:748)
[info] <JMH had finished, but forked VM did not exit, are there stray running threads? Waiting 19 seconds more...>

Reply:
Looks like something is broken in that benchmark, `flow`, which it expects to run is never assigned. Missing `flow = ` on Line 73 I guess.

Reply:
Looks like 5 iterations for warmup was a bit low (the `-w 5` parameter), so perhaps do 10 there instead, so that you see that the number becomes more stable before the actual bench iterations.

Reply:
@johanandren no luck(
https://gist.github.com/yeryomenkom/2bb83620fb4e6423e87bf2ec39b7eb41

Reply:
Hmm, looks like the exact same NPE, I think JMH can be a bit finicky with recompilation so a clean compile inbetween can help (did work locally for me without clean though), given that you did the last change, so that Source.... that is in there in the last line of the setup now should be assigned to the `flow` field, since that is the stream the actual test is running.